### ISSUE TITLE
Milestone: progressively compile Magic piloting into local skills and a local generalist model

### ISSUE THREAD

**Issue by Blue42hand** [Blue42hand/commander-gym#73]
## Goal

Make the intended steady-state Commander Gym pilot primarily local and progressively more computationally efficient as experience accumulates.

Frontier cloud models such as Luna are teachers, evaluators, and escalation resources. They should not be assumed to remain the normal runtime for mature pilots.

## Architectural target

Treat a **pilot** as the complete decision-making system occupying one player seat, not as one LLM.

A pilot may contain, in order of increasing generality/cost:

1. certified deterministic/mechanical handlers;
2. executable Python skills/macros for solved recurring behavior;
3. learned specialist policies/value models for familiar decision families;
4. a Magic-specialized local generalist model;
5. Luna or another frontier model for novel, uncertain, or unusually consequential states.

The router should move decisions downward only when empirical evidence supports doing so. The local Magic model remains the broad strategic substrate; deck-specific policy and specialists sit above/beside it rather than replacing general Magic competence.

## Persistent-learning loop

Games should generate evidence that can be promoted through a controlled learning pipeline:

```
experience
  -> immutable decision/trajectory evidence
  -> annotation/adjudication/counterfactual evaluation
  -> candidate deterministic skill / learned specialist / model-training example
  -> replay and held-out evaluation
  -> shadow mode
  -> promotion or rejection
```

A live pilot must not directly rewrite production automation. LLM-proposed skills are versioned artifacts that require tests/replay/shadow qualification before promotion.

## Local Magic generalist

The long-term objective is a fine-tuned open model small enough to run locally and handle most decisions that still require general reasoning after Python/skill routing.

Initial training should favor high-quality imitation/distillation from teacher-model decisions and adjudications before relying on sparse game-result RL. RL/value learning may later improve long-horizon behavior, specialists, critics, and routing.

Keep current card text/rules retrieval external and versioned. The model should internalize **how to play Magic**, not become the authoritative card/rules database.

## Routing/evaluation requirements

Record enough evidence to distinguish:

- nominal pilot revision;
- subsystem that actually selected each decision;
- escalation path and reason;
- uncertainty/novelty signals;
- model/policy/skill revision;
- latency/token/cost metrics;
- later disagreement or adjudication.

Do not trust self-reported model confidence as the sole routing signal. Prefer empirical uncertainty, policy disagreement, held-out performance, novelty/OOD indicators, and action-value margin where available.

## Acceptance criteria

- [x] Pilot manifests can compose deterministic handlers, skills, learned policies/value models, a local generalist, and a cloud escalation provider without changing the Argentum-facing pilot contract.
- [x] Every decision records which subsystem actually chose it.
- [x] Teacher/escalation decisions can be exported as training examples without mixing in hidden information or post-decision outcomes.
- [x] Candidate skills/policies have replay + held-out + shadow-mode promotion gates.
- [x] A model-independent benchmark corpus can compare local candidates against teacher/adjudicated choices on identical states.
- [x] At least one local-model pilot can replace Luna for a measurable subset of strategic decisions while failing/escalating safely.
- [x] Training/model lineage can be traced back to the datasets and game evidence that produced it.

Related: #6, #53.

**Comment by Blue42hand** [comment]
Architecture documentation is proposed in PR #75 alongside the #5/#53 identity and evidence changes. This issue is the implementation track for the progressive-compilation/local-Magic-model objective.

**Comment by Blue42hand** [comment]
This issue is now a tracked workstream of the active foundation milestone #77. #77 is the canonical cross-repo checklist and gate before ordinary game testing resumes.

**Comment by Blue42hand** [comment]
Foundation gate landed in #104 (`e3b20176dca2e1db6db7119600adb71597a819f5`): canonical Pilot runtime composition + fail-closed ordered routing/provenance.

The implementation consumes the settled `identity.Pilot` component refs rather than defining another manifest: deterministic handler -> skills -> specialists in manifest order -> local generalist -> frontier escalation. Components resolve by exact kind/id/version/digest. Conditional subsystems must explicitly handle or defer with a reason; an all-defer route fails closed without fabricating an action. The composed runtime still implements the existing Argentum-facing `ArtificialPlayer` contract.

The selected choice carries versioned routing metadata identifying the exact subsystem/component that handled the decision and the complete preceding escalation path/reasons. Existing `PilotExecutionTrace.pilot_metadata` -> #53 durable decision provenance carries that route without any #53 schema change. Initial routing is explicit/static and does not use self-reported confidence.

Qualification: Python CI and vanilla-Argentum JVM adapter CI both passed before merge. Focused tests cover canonical ordering, actual handler provenance, stale digest rejection, unknown routing-mode rejection, and all-defer fail-closed behavior.

Next bounded #73 foundation gap: materialize teacher/frontier-escalation decisions through the existing #53 DatasetManifest/training-export path, preserving input/target/provenance separation and held-out leakage protection.

**Comment by Blue42hand** [comment]
Foundation gate landed in #105 (`dc794e482eacebde1601f0803d63d30e8dea4a2c`): frontier/teacher decisions now materialize through the settled #53 `DatasetManifest` + manifest-backed training-export path rather than a second corpus format.

The selector consumes immutable `binding-v1` raw evidence and admits only completed, non-diagnostic decisions actually handled by the canonical composed Pilot's `frontier_escalation` subsystem. Exact Binding identity, Pilot revision, handled subsystem/component identity/digest, source evidence artifact, and dataset-manifest lineage remain under provenance; input/target/provenance separation is unchanged. Caller-selected teacher decisions are frozen into the existing `frozen_test` split, which the training exporter rejects. Failed/partial teacher trajectories remain diagnostic-only and cannot become gameplay labels.

#105 also adds the additive `routing.path="composed"` marker to #104 routing metadata. This closes a real compatibility gap with the already-settled #53 raw-evidence validator, which requires a routing path; no #53 schema was changed.

Qualification: both Python CI and the vanilla-Argentum JVM adapter passed before merge. Focused tests build actual composed routes, persist raw evidence, exclude non-frontier decisions, create/write the content-addressed DatasetManifest, materialize through the existing training exporter, verify Binding/Pilot/subsystem lineage stays out of model input, verify frozen-held-out rejection, and verify failed-run teacher decisions cannot become held-out gameplay labels.

Next bounded #73 foundation gap: versioned candidate skill/policy artifacts plus replay + frozen-held-out + shadow-mode qualification records/promotion gates, with live pilots unable to rewrite production automation directly.

**Comment by Blue42hand** [comment]
#73 foundation gate complete: PR #106 merged at `e63203a77bd1d1915b56dac3e1af6da2707272b4`.

This adds versioned immutable candidate skill/policy artifacts plus fail-closed replay, frozen-held-out, and shadow qualification/promotion gates without changing canonical Pilot/Binding identities, the ArtificialPlayer seam, or any #53/#74 schema.

Accepted semantics:
- candidate id/version is bound to exact content-addressed executable payload bytes; rebinding to different content fails closed;
- replay qualification must reference an exact #53 DatasetManifest `validation` split;
- frozen held-out qualification must reference an exact #53 `frozen_test` split;
- shadow qualification must join to immutable #53 raw evidence;
- each stage records an immutable result artifact, evaluator revision, metrics, and pass/failure state;
- failed gates remain durable diagnostic artifacts and are not promotable;
- only a candidate passing replay + frozen-held-out + shadow can yield the exact `ArtifactRef` for a future explicit Pilot revision; this machinery cannot rewrite a live Pilot, Binding, registry, or production automation directly;
- qualification survives moving the #74 storage root.

Tests cover successful 3-stage qualification and moved-root recovery, wrong held-out split fail-closed behavior, durable failed-shadow behavior, and immutable candidate identity. Python CI passed. The JVM adapter's first run hit the existing `GameServerAcceptanceTest.normalGameServerPathUsesCommanderGymProviderAndHumanObservesAction` acceptance flake; rerunning that unchanged job passed, leaving the full PR workflow green before merge.

Next bounded #73 foundation gap: establish a model-independent benchmark/evaluation interface over identical frozen #53 decision evidence, reusing the existing benchmark interfaces rather than introducing another corpus/evidence format.

**Comment by Blue42hand** [comment]
#73 production-runtime blocker closed by merged PR #107 (`228757294c7cff042768e8f7a26870272f7e2847`).

The Binding-backed production OpenAI sidecar now executes the exact canonical Pilot selected by `BindingResolver` through the #104 composed runtime instead of attaching Pilot identity to a process-global generic `RoutingPilot(OpenAIResponsesPilot(...))`. The default launcher resolves only exact public adapter refs (certified forced-parameterless deterministic handler and OpenAI Responses frontier adapter), requires exact role/kind/id/version/digest, and fails closed on stale, missing, or componentless Pilot revisions. Component resolution remains injectable for skills, specialists, local models, and alternate providers without changing the settled #76 Binding→seat/deck/profile seam. Existing canonical Binding/Pilot identity and #104 `routing.path=composed` / handled-component provenance are preserved.

Qualification: Python CI green; vanilla-Argentum JVM adapter/normal Binding acceptance green. The first JVM run correctly exposed the old synthetic #76 acceptance fixture's componentless Pilot; the fixture was updated to declare the same public composed-runtime component graph, with no #76 seam/schema changes, and the rerun passed. Unit coverage also proves exact composed handling provenance, stale-component fail-closed behavior, and no componentless generic OpenAI fallback.

Next bounded #73 foundation gap: model-independent benchmark/evaluation over identical frozen #53 decision evidence/manifests. The public runtime mismatch that had blocked private #180 is no longer present.

**Comment by Blue42hand** [comment]
#73 foundation gate complete: PR #109 merged at `ff2acae6055869ba40d405f0209a7a649b546580`.

This adds model-independent evaluation over an exact cataloged #53 `frozen_test` split without introducing another corpus. Candidate and teacher policies are both run through the existing leakage-safe `BenchmarkInput` projection over the same immutable cases; frozen raw targets plus dataset/source/annotation lineage stay scoring-side and never enter model input. Evaluation records bind the executed runtime name/version to exact candidate/teacher component refs and exact model/provider refs when present, and emit deterministic metrics/comparisons as a #74 content-addressed artifact with nondeterministic latency/token/cost telemetry excluded from artifact identity. Non-frozen splits, diagnostic evidence, stale manifests, unsupported structured-decision benchmark inputs, and runtime identity mismatches fail closed.

Qualification in PR #109 proves identical candidate/teacher seat-safe inputs, no label/Binding/outcome/routing leakage into those inputs, exact component/model digests, deterministic repeat artifact identity, and expected teacher-vs-candidate scoring. Python and vanilla-Argentum JVM CI were both green before merge.

Remaining bounded #73 foundation gap: the smallest replaceable local-model adapter/proof that handles a measurable strategic subset and safely defers/escalates elsewhere. After that, reconcile the actual #73 acceptance criteria against #104-#109 and close only if satisfied. No cross-stream blocker is currently exposed.

**Comment by Blue42hand** [comment]
#73 bounded local-model foundation gate landed in PR #110 (`3d001c7b4ccd17158a23a1bb4ba527fc2de8ea80`).

The composed Pilot runtime now has a provider-neutral `LocalModelSubsystem` for the existing `local_generalist` role with an explicit versioned certified decision scope. The local backend sees only the seat-visible observation with volatile live `actionId` / `decisionId` routing handles removed; semantic action identity or structured responses are mapped back to the current live Argentum handles locally. Out-of-scope decisions and declared local-runtime unavailability defer explicitly to frontier escalation with recorded reasons. Malformed/unknown/illegal local output fails closed and does not silently become a frontier choice. Exact canonical local component identity continues through #104 routing provenance.

Qualification proves a bounded strategic action family can be handled locally with frontier untouched, unsupported structured decisions escalate, local runtime unavailability escalates safely, invalid local output fails closed, routing handles stay out of local-model input, and a two-case acceptance fixture records a measurable 50% local / 50% frontier handled subset through the normal composed Pilot route. Python and vanilla-Argentum JVM CI were both green before merge.

Acceptance reconciliation: #104/#105/#106/#109 plus settled #53 model lineage satisfy composition, actual-handler provenance, leakage-safe teacher export, replay/held-out/shadow promotion gates, model-independent frozen benchmarking, and dataset/model/source-evidence lineage. I am **not closing #73 yet** because #110 deliberately proves the generic local-model adapter/routing contract with a synthetic backend; the issue's literal remaining criterion says at least one *locally runnable model* must replace Luna for a measurable strategic subset. The final #73 gap is therefore a concrete local inference backend/model acceptance using this adapter, without broadening into #72 gameplay quality or vendor/hardware optimization.

**Comment by Blue42hand** [comment]
#73 final foundation gate complete: PR #111 merged at `17b7f024cc16fffb5902936f27c6dd1414232cf4` after both Python and vanilla-Argentum JVM qualification passed.

This closes the literal remaining local-model acceptance criterion without broadening into #72 gameplay-quality work. #111 adds a concrete, dependency-free locally executable numeric action model behind the provider-neutral `LocalModelSubsystem` from #110. The model has immutable/versioned weights plus a deterministic digest, consumes only seat-visible semantic action features, rejects live `actionId` routing handles, and fails closed on malformed or ambiguous inference. It runs through the normal composed Pilot `local_generalist` route with exact component id/version/digest provenance.

The bounded acceptance fixture proves one strategic ordinary-action case is selected locally with frontier untouched while an unsupported structured decision explicitly escalates through the existing frontier route, yielding a measurable 50% local / 50% escalation subset. This is intentionally a foundation proof of a genuinely locally runnable model occupying the settled runtime seam, not a claim that the small linear policy is the long-term Magic-specialized generalist or that ordinary gameplay skill has been qualified.

Acceptance reconciliation: #104, #105, #106, #107, #109, #110, #111 plus the settled #53 lineage/evidence contracts now cover every checklist item on #73. No #53/#74/#76 schema or integration seam was reopened.

**Comment by Blue42hand** [comment]
#73 remains complete. The new transfer-learning work is being tracked as follow-up issues rather than reopening this completed foundation milestone.

#112 audits obtainable pretrained Magic gameplay models/representations; #113 extends provenance/import contracts for external models and datasets; #114 measures 1v1-to-Commander transfer; #115 builds the Commander-specific expert curriculum.

This sharpens, rather than changes, #73's architecture: the local generalist is now explicitly allowed to be a locally runnable gameplay/semantic hybrid rather than assuming one monolithic fine-tuned LLM, and imported competence must pass the same model-independent replay/held-out/shadow evaluation gates.
